### PR TITLE
[FIRRTL] Change Port Direction attribute from an APInt to a DenseArray.

### PR DESCRIPTION
**Pull request by darthscsi**
APInt requires an allocation to start large values.  Using apint for port directions requires copying the apint when inspecting port directions, thus introducing huge amount of copies throughout the code.  On large designs just this allocation overhead amounted to over 25% of the time spent (plus copy times).  DenseArrays are less space efficient (by 8x as used), but can be accessed by reference, removing these allocations.

Overall improvement is about 35% - 40%.

### CODE REVIEW

File: include/circt/Dialect/FIRRTL/FIRRTLAttributes.h [hunk 1]
```diff
@@ -33,22 +33,35 @@ namespace direction {
...
+static inline StringRef toString(Direction direction) {
   return direction == Direction::In ? "in" : "out";
 }
 
+static inline StringRef toString(bool direction) {
+  return toString(get(direction));
+}
+
+/// Return a \p IntegerAttr containing the packed representation of an array
+/// of directions.
```
Comment (seldridge) [inline include/circt/Dialect/FIRRTL/FIRRTLAttributes.h:53]:
```suggestion
/// Return a \p DenseBoolArrayAttr containing the packed representation of an array
/// of directions.
```

Ditto for other comments which are now wrong below.

File: include/circt/Dialect/FIRRTL/FIRRTLStructure.td [hunk 2]
```diff
@@ -130,6 +130,7 @@ def FModuleOp : FIRRTLModuleLike<"module", [SingleBlock, NoTerminator]> {
   }];
   let arguments =
       (ins ConventionAttr:$convention,
+           DenseBoolArrayAttr:$portDirections,
```
Comment (seldridge) [inline include/circt/Dialect/FIRRTL/FIRRTLStructure.td:133]: It's weird that this _wasn't_ in the argument list before.

File: lib/Conversion/ExportVerilog/ExportVerilog.cpp [hunk 3]
```diff
@@ -221,7 +221,8 @@ static void emitZeroWidthIndexingValue(PPS &os) {
 
 /// Return the verilog name of the port for the module.
 static StringRef getPortVerilogName(Operation *module, size_t portArgNum) {
-  return cast<HWModuleLike>(module).getPort(portArgNum).getVerilogName();
+  auto hml = cast<HWModuleLike>(module);
+  return hml.getPort(portArgNum).getVerilogName();
```
Comment (seldridge) [inline lib/Conversion/ExportVerilog/ExportVerilog.cpp:225]: Nit: unrelated change? Also, IMO inlining single-use seems better.

File: lib/Conversion/HandshakeToHW/HandshakeToHW.cpp [hunk 4]
```diff
@@ -290,7 +290,7 @@ portToFieldInfo(llvm::ArrayRef<hw::PortInfo> portInfo) {
 // Convert any handshake.extmemory operations and the top-level I/O
 // associated with these.
 static LogicalResult convertExtMemoryOps(HWModuleOp mod) {
-  auto ports = mod.getPortList();
+  //  auto ports = mod.getPortList();
```
Comment (seldridge) [inline lib/Conversion/HandshakeToHW/HandshakeToHW.cpp:293]: Delete if not used.

File: lib/Dialect/FIRRTL/FIRRTLAttributes.cpp [hunk 5]
```diff
@@ -34,26 +34,27 @@ Direction direction::flip(Direction direction) {
...
+mlir::DenseBoolArrayAttr
+direction::packAttribute(MLIRContext *context, ArrayRef<Direction> directions) {
   // Pack the array of directions into an APInt.  Input is zero, output is one.
-  auto size = directions.size();
-  APInt portDirections(size, 0);
-  for (size_t i = 0; i != size; ++i)
-    if (directions[i] == Direction::Out)
-      portDirections.setBit(i);
-  return IntegerAttr::get(IntegerType::get(context, size), portDirections);
+  SmallVector<bool> dirs;
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/FIRRTLAttributes.cpp:40]: Doing a resize (as is done in a later method) would avoid repeated multiple allocations.

File: lib/Dialect/FIRRTL/FIRRTLAttributes.cpp [hunk 6]
```diff
@@ -34,26 +34,27 @@ Direction direction::flip(Direction direction) {
...
+
+SmallVector<Direction>
+direction::unpackAttribute(mlir::DenseBoolArrayAttr directions) {
   SmallVector<Direction> result;
-  result.reserve(size);
-  for (size_t i = 0; i != size; ++i)
-    result.push_back(direction::get(value[i]));
+  result.reserve(directions.size());
+  for (size_t i = 0ULL, e = directions.size(); i != e; ++i)
+    result.push_back(direction::get(directions[i]));
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/FIRRTLAttributes.cpp:57]: The `size_t` index is likely unnecessary now that this isn't an `IntegerAttr`. This is cleaner as:

```
for (auto d : directions) ...
```

Or similar.

File: lib/Dialect/FIRRTL/Transforms/InferResets.cpp [hunk 7]
```diff
@@ -857,7 +857,6 @@ void InferResetsPass::traceResets(InstanceOp inst) {
   LLVM_DEBUG(llvm::dbgs() << "Visiting instance " << inst.getName() << "\n");
 
   // Establish a connection between the instance ports and module ports.
-  auto dirs = module.getPortDirections();
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/Transforms/InferResets.cpp:860]: How is there just dead code laying around? 😖

File: test/Dialect/FIRRTL/errors.mlir [hunk 8]
```diff
@@ -35,7 +35,7 @@ firrtl.circuit "" {
 
 firrtl.circuit "foo" {
 // expected-error @+1 {{requires 1 port directions}}
-firrtl.module @foo(in %a : !firrtl.uint<1>) attributes {portDirections = 3 : i2} {}
+firrtl.module @foo(in %a : !firrtl.uint<1>) attributes {portDirections = array<i1: true, true>} {}
```
Comment (seldridge) [inline test/Dialect/FIRRTL/errors.mlir:38]: FWIW, this is easier to read now. 🙃

**Review by seldridge — APPROVED** [review]
Nits throughout. Looks good otherwise.

File: include/circt/Dialect/FIRRTL/FIRRTLStructure.td (hunk 2 above)
Comment (darthscsi) [reply include/circt/Dialect/FIRRTL/FIRRTLStructure.td:133]: Nothing was, basically.  Added most of the things last week, but apparently missed this one.

File: lib/Dialect/FIRRTL/FIRRTLAttributes.cpp (hunk 6 above)
Comment (darthscsi) [reply lib/Dialect/FIRRTL/FIRRTLAttributes.cpp:57]: no, because DenseArrayRef don't have begin and end!

File: include/circt/Dialect/FIRRTL/FIRRTLAttributes.h [hunk 9]
```diff
@@ -33,22 +33,35 @@ namespace direction {
...
+static inline StringRef toString(bool direction) {
+  return toString(get(direction));
+}
+
+/// Return a \p DenseBoolArrayAttr containing the packed representation of an array
+/// of directions.
+mlir::DenseBoolArrayAttr packAttribute(MLIRContext *context,
+                                       ArrayRef<Direction> directions);
+
 /// Return a \p IntegerAttr containing the packed representation of an array
```
Comment (SpriteOvO) [inline include/circt/Dialect/FIRRTL/FIRRTLAttributes.h:57]:
```suggestion
/// Return a \p DenseBoolArrayAttr containing the packed representation of an array
```
The comment should be updated.

File: lib/Conversion/ExportVerilog/ExportVerilog.cpp (hunk 3 above)
Comment (darthscsi) [reply lib/Conversion/ExportVerilog/ExportVerilog.cpp:225]: Unrelated, but matches other code adjacent.  Will be rewritten soon.
